File: CafeApi/Models/Order.cs
using CafeApi.Enums;
using CafeApi.Interfaces;

namespace CafeApi.Models;

public class Order : IEntity
{
    public int Id { get; private set; }
    public int CustomerId { get; private set; }
    public Customer Customer { get; private set; } = null!;
    public OrderStatus Status { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public decimal FinalTotal { get; private set; }
    public ICollection<OrderItem> Items { get; private set; } = [];

    public decimal Total => Items.Sum(x => x.Total);

    private Order()
    {
    }

    public static Order Create(
        int customerId,
        ICollection<OrderItem> items
    )
    {
        var order = new Order
        {
            CustomerId = customerId,
            Status = OrderStatus.Pending,
            CreatedAt = DateTime.UtcNow,
            Items = items
        };
        order.FinalTotal = order.Total;
        return order;
    }

    public void ApplyPromotion(CustomerPromotion customerPromotion)
    {
        if (Status != OrderStatus.Pending)
            throw new InvalidOperationException("Promotion can only be applied to pending orders");

        FinalTotal = customerPromotion.Promotion.CalculateDiscountedPrice(Total);
        customerPromotion.MarkAsUsed(this);
    }

    public void UpdateStatus(OrderStatus status) => Status = status;
}

File: CafeUI/src/types/order.ts
import type {OrderStatus} from "./enums/enums.ts";

export interface CreateOrderDto {
  customerId: number
  items: OrderItemDto[]
}

export interface OrderItemDto {
  menuItemId: number
  quantity: number
}

export interface OrderItemResponseDto {
  menuItemName: string
  quantity: number
  unitPrice: number
}

export interface OrderResponseDto {
  id: number
  customerName: string
  customerId: number
  status: keyof typeof OrderStatus
  createdAt: string
  total: number
  finalTotal: number
  items: OrderItemResponseDto[]
}